### ISSUE TITLE
2.5/3D input baselines

### ISSUE THREAD

**Issue by someone-TJ** [astra-vision/MonoScene#66]
Hello，Why do the metrics in your article comparing the models differ so much from the metrics shown in their own papers, such as AICNet is 33% and 3DSketch is 41.1% in their own papers?

**Comment by anhquancao** [comment]
Hi @someone-TJ, because in our setting, the 3D input is computed from the depth predicted from the RGB image, not the ground-truth depth.

**Comment by someone-TJ** [comment]
Well, I was referring to the 2.5/3D input baseline, not the RGB input baseline, which is 23.8% for AICNet and 29.2% for 3DSketch in the 2.5/3D input baseline, but 33% and 41.1% respectively in their own paper.

**Comment by anhquancao** [comment]
I don't remember everything exactly but the reasons are roughly:
1. For SSC task with 3D input, we normally only train and evaluate the performance on the occluded and observed surface.  You can refer to [this survey Fig. 2](https://arxiv.org/pdf/2103.07466.pdf) . The observed free and the observed surface are extracted directly from the 3D input. In our case, since we need to evaluate on the full scene. 
2. For these papers, they directly use the 3D ground-truth to extract the occluded area, **not** the input depth.
3. However, there is a big **mismatch** between the 3D ground-truth and the recorded depth (missing depth/wrong annotation).  It thus leads to a big performance drop when using the input depth directly to extract the observed free and observed surface. This subsequently leads to the drop in the final performance on the full scene.

**Comment by someone-TJ** [comment]
I have the following questions:
1. in the article you mention 13 classes for NYUv2, including a free and a unknown, but in [monoscene/data/NYU/preprocess.py](https://github.com/astra-vision/MonoScene/blob/master/monoscene/data/NYU/preprocess.py) the [seg_class_map ](https://github.com/astra-vision/MonoScene/blob/5239dfdbc33a5ae5d7bc895b3edadcc4ca14ce53/monoscene/data/NYU/preprocess.py#L11)only defines 12 tags, 0-11; does this mean that the free class is being evaluated, but the unknown class is not
2.I don't quite understand your second point, the reason why AICnet and 3DSketch have lower metrics in your evaluation criteria than mentioned in their own articles is because  they don't evaluate the outside view mentioned in [this survey Fig. 2](https://arxiv.org/pdf/2103.07466.pdf)？ but your evaluation needs to include the outside view？

**Comment by anhquancao** [comment]
1. For NYUv2, there is no unknown class. Only SemKITTI has unknown class which contains the voxels that are not observed by any lidar rays in the sequence.
2. I mean AICnet and 3DSketch train and evaluate only on the green and blue area in the image. The red area (semantic + occ) and the green surface (occ) are assumed to be extracted directly from the input depth. 
![image](https://github.com/astra-vision/MonoScene/assets/9668858/49332fd5-fb0e-4a8d-bb1d-c2c72a98fa69)

However, since there are mismatch between the 3D ground-truth and the input depth (missing depth/wrong annotation), the volume extracted directly from the input depth has a lot of error. Thus, AICnet and 3DSketch  instead use the 3D ground-truth to get the red area (semantic + occ) and the green surface (occ), this is not possible in camera-only setting.

Therefore, there are two ways to compare:
- Extract the the red area (semantic + occ) and the green surface (occ) from the input depth and only optimize the rest of the volume.
- Optimize directly the full volume. This one give best results and are reported in the paper.

**Comment by someone-TJ** [comment]
if I wanna train and evaluate only on the green and blue area in the image，how do I modify the metric code？

**Comment by anhquancao** [comment]
You need to find the voxels in the red area from the input depth by doing ray marching.

**Comment by someone-TJ** [comment]
Is it possible to use [sscMetrics.py](https://github.com/waterljwant/SSC/blob/master/sscMetrics.py)  directly？

**Comment by anhquancao** [comment]
The problem is that you need to compute the mask to mask out the unrelevant area. For NYUv2, maybe you can get the mask from the 3DSketch repo https://github.com/charlesCXK/TorchSSC
For SemKITTI, I think they have the occluded annotation stored in .occluded files.

**Comment by someone-TJ** [comment]
Thank you for the suggestion. Would Monscene still be better than AICnet if only the observed surfaces and obscured voxels were evaluated for metrics?

**Comment by anhquancao** [comment]
I would guess yes, since the gap on full scene performance is quite big (+8 mIoU compared to AICNet). But you should reevaluate to confirm. The mask for NYUv2 can be downloaded from 3DSketch repo https://github.com/charlesCXK/TorchSSC. Note that this mask is obtained from the 3D ground-truth, not the input depth.

**Comment by someone-TJ** [comment]
How do I get the Iou for each class in the 2.5D/3D input baseline of the NYUv2 test set？ I only found the SC Iou and the SSC mIou

**Comment by someone-TJ** [comment]
well,I mean the per class IoU of LMSCnet,AICnet,and 3dSetch in the 2.5D/3D baseline

**Comment by anhquancao** [comment]
Yes, it's the code for the baselines but I didn't keep the checkpoints.

**Comment by someone-TJ** [comment]
> The problem is that you need to compute the mask to mask out the unrelevant area. For NYUv2, maybe you can get the mask from the 3DSketch repo https://github.com/charlesCXK/TorchSSC For SemKITTI, I think they have the occluded annotation stored in .occluded files.

I design the mask simplify,but I'm not sure it's fair. I set mask = [gt == 255 & gt == 0] = 0, and I get the voxel position which need evaluate by make mask == 1, Equivalent to the case where I only evaluate 11 categories (excluding null and 255) in the occupied area. finally I get 37.5% mIoU.

**Comment by anhquancao** [comment]
Can you detail more? 37.5% mIoU is performance of which model, which dataset? What is the training procedure?
For NYUv2, the mask is stored in[ label_weight variable](https://github.com/charlesCXK/TorchSSC/blob/master/model/sketch.nyu/eval.py#L99) .

**Comment by someone-TJ** [comment]
After I modifed the evaluation, I retrained monoscene on NYUv2. and I keep your training setting and code. My result is following as：
Precision=100.0000, Recall=73.3334, IoU=73.3334
class IoU: ['empty', 'ceiling', 'floor', 'wall', 'window', 'chair', 'bed', 'sofa', 'table', 'tvs', 'furn', 'objs'], 
0.0000,  23.0571,  93.7877,  26.3159,  26.2221,  28.8364,  62.1524,  54.2303,  19.8685,  17.5469,  42.5319,  18.9657, 
mIoU=37.5923
I didn't realize the mask actually is label_weight variable. and I am confused with some process procedure in 3dsketch like 
[k = (gt >= 0) & (gt < n_cl)  # exclude 255](https://github.com/charlesCXK/TorchSSC/blob/2fd21ad25af92cd9f9ad28de3c4bc897c0ae8b43/model/sketch.nyu/eval.py#L74)
It seems that empty is still evaluated in 3dsketch instead of only evaluating the occlude and the surface

**Comment by anhquancao** [comment]
Yes, you still have empty space in the occluded space. You should use their label_weight mask. 
It's weird that the precision is 100. You still need to predict the empty voxels that are in the occluded area.

**Comment by someone-TJ** [comment]
I will use their label_weight mask and try it again. by the way, is the evaluation differenyt between AICnet and 3dsketch? I found the definiton is similar with mask in AICnet,      
[def get_nonempty2(voxels, target, encoding):  # Get none empty from depth voxels](https://github.com/waterljwant/SSC/blob/0a347cac34bb8db9579f866e16f4192636646bf1/dataloaders/dataloader.py#L377)
If the same does it mean that I can also use the evaluation in AICnet to evaluate

**Comment by anhquancao** [comment]
I am not sure. I don't remember exactly. You should double check it.

**Comment by anhquancao** [comment]
nonsurface is not use, you can ignore it. 
Yes, you can do mask = mask & nonempty. Please make sure to extract the correct mask from label_weight as in 3DSketch.

**Comment by someone-TJ** [comment]
My reproduce result is following as:
**Precision=76.9471, Recall=65.0203, IoU=54.4195
class IoU: ['empty', 'ceiling', 'floor', 'wall', 'window', 'chair', 'bed', 'sofa', 'table', 'tvs', 'furn', 'objs'], 
61.9469,  15.3907,  93.8814,  16.8636,  19.1930,  19.8496,  55.1806,  43.4497,  19.3723,  19.4189,  33.0244,  18.1927, 
mIoU=32.1652**
It is lower than the 34.1% mentioned in the original AICnet paper. But in 2.5/3D input baselines of your paper, monoscene is actually higher than AICnet 23.9%. It's weird

**Comment by anhquancao** [comment]
So it's normal,  34.1% of AICNet is with true depth as input not only RGB.

**Comment by someone-TJ** [comment]
well, AICnet with true depth and rgb as input is just 23.8 % if evalute on full scene. I am confused why it's lower than monoscene on full scene.

**Comment by anhquancao** [comment]
23.8 % AICnet is not only obtained from reevaluation. I retrained the AICNet on full scene.

**Comment by someone-TJ** [comment]
This is indeed the case. I checked the source code of AICnet and 3dsketch, and 3dsketch does use mask to filter the observed and occluded regions during training as well, but in AICnet, it seems to use nonempty only during testing 
[if self.istest:](https://github.com/waterljwant/SSC/blob/0a347cac34bb8db9579f866e16f4192636646bf1/dataloaders/dataloader.py#L70)

**Comment by anhquancao** [comment]
Yes, maybe AICNet trains on the full scene. Btw, can you verify if you got 34.1% when running AICNet code? I remember I can't reproduce their value.

**Comment by someone-TJ** [comment]
Sorry, I misremembered, the result shown in the original AICnet article was 33.3%. I tried to reproduce AICnet but it didn't work , when I run the training program, the progress almost stuck , I don't know what is the problem. so you reproduce how much of their work can be achieved

**Comment by anhquancao** [comment]
I checked. I got 53.65% IoU and 28.23% mIoU for AICNet.

**Comment by someone-TJ** [comment]
Fine,I will train monoscene on the "visible surface and occluded space" again so that can compare with other SSC model  which evaluated on the "visible surface and occluded space". Thank you for your patient help

**Comment by anhquancao** [comment]
You are welcome. Good luck with your research!
